README.md
```

             :::::::::::::::::::
             :::   _______   :: ________ ____.  ______.    .____.
             :::  .\   _  \____.\  _.  /.\   |_\\     |----|    |
             :::  |    |  .    |   \|   |    _/---    |    |    |_
             :::  |    |\_|    |    :   |    |    \___     |      \
             :::  |    |  |____|________|_____    | //____/|______/ H7
         - - ::: -`----' --- ... ----------- `----' ---------------- -- -
             :::             :::
             :::::::::::::::::::

```

## Description

Motyl is an opinionated blog-aware static site generator written in Ruby.
It uses Mustache as templating system, and all content is written in Markdown.

For the record, motyl means butterfly in Polish.

## Features

- Small and easy to understand codebase
- Minimal dependencies (only three gems)
- Pages and posts written in Markdown
- Templates are logic-less and use Mustache
- Support for multiple categories per post
- Syntax highlighting (using Rouge)
- Customizable URLs (constructed from filename)
- Atom feed generator

## Requirements

### Ruby modules

Motyl requires the following Ruby modules:

- kramdown
- mustache
- rouge

Installing dependencies via gem:

	gem install kramdown mustache rouge

Alternatively, those modules can be installed directly via the operating
system's package manager.

## Configuration

The 'examples' directory contains a sample site which can be used as a
starting point.

### Installing a theme

Clone the [Chrysalide](https://github.com/fcambus/chrysalide) theme repository
and place it in the `themes` directory.

## Usage

Simply run `motyl` to build the site, it will generate posts and pages into
the `public` directory, and will also copy static assets.

## License

Motyl is released under the BSD 2-Clause license. See `LICENSE` file
for details.

## Author

Motyl is developed by Frederic Cambus.

- Site: https://www.cambus.net

## Resources

GitHub: https://github.com/fcambus/motyl


ChangeLog
Motyl 1.0.3 (2022-07-28)

- Fix an off-by-one error, we want 20 posts in the Atom feed
- Raise the default to 30 posts in the Atom feed
- Remove unused feedItems directive from motyl.conf
- Add Motyl ASCii logo header (Thanks H7!)
- Add SPDX short license identifier in source file



Motyl 1.0.2 (2019-05-07)

- Add require 'date', which allows Motyl to run with Ruby 2.5+



Motyl 1.0.1 (2018-10-18)

- Enforce UTF-8 character encoding
- Lint code and wrap lines longer than 80 characters



Motyl 1.0.0 (2018-04-28)

- Initial release


src/motyl.lua
#!/usr/bin/env lua
--[[
###############################################################################
#                                                                             #
# Motyl                                                                       #
# Copyright (c) 2016-2018, Frederic Cambus                                    #
# https://github.com/fcambus/motyl                                            #
#                                                                             #
# Created: 2016-02-16                                                         #
# Last Updated: 2018-03-04                                                    #
#                                                                             #
# Motyl is released under the BSD 2-Clause license.                           #
# See LICENSE file for details.                                               #
#                                                                             #
###############################################################################
]]--

local lfs = require "lfs"
local lyaml = require "lyaml"
local lunamark = require "lunamark"
local lustache = require "lustache"

-- Read data from file
local function readFile(path)
	local file = assert(io.open(path, "rb"))

	local data = file:read "*all"
	file:close()

	return data
end

-- Write data to file
local function writeFile(path, data)
	local file = assert(io.open(path, "wb"))

	file:write(data)
	file:close()
end

-- Load YAML from file
local function loadYAML(path)
	return lyaml.load(readFile(path))
end

-- Load and process Markdown file
local function loadMD(path)
	local writer = lunamark.writer.html.new()
	local parse = lunamark.reader.markdown.new(writer, { fenced_code_blocks = true })
	return parse(readFile(path))
end

-- Sorting function to sort posts by date
local function sortDates(a,b)
	return a.date > b.date
end

-- Display status message
local function status(message)
	print("[" .. os.date("%X") .. "] " .. message)
end

-- Loading configuration
local data = {}
data.version = "Motyl 1.00"
data.updated = os.date("%Y-%m-%dT%XZ")
data.site = loadYAML("motyl.conf")
data.site.feed = {}
data.site.posts = {}
data.site.categories = {}

-- Loading templates
local templates = {
	categories = readFile("themes/templates/categories.mustache"),
	header = readFile("themes/templates/header.mustache"),
	atom = readFile("themes/templates/atom.mustache"),
	pages = readFile("themes/templates/page.mustache"),
	posts = readFile("themes/templates/post.mustache"),
	footer = readFile("themes/templates/footer.mustache")
}

local function render(directory)
	for file in lfs.dir(directory) do
		if file ~= "." and file ~= ".." then
			local extension = file:match "[^.]+$"

			if extension == "md" then
				local path = file:match "(.*).md$"
				data.page = loadYAML(directory .. "/" .. path .. ".yaml")
				data.page.content = lustache:render(loadMD(directory .. "/" .. file), data)
				if data.page.url == nil then
					data.page.url = path .. "/"
				end

				status("Rendering " .. data.page.url)

				if directory == "posts" then
					local year, month, day, hour, min = data.page.date:match("(%d+)%-(%d+)%-(%d+) (%d+)%:(%d+)")
					data.page.datetime = os.date("%Y-%m-%dT%XZ", os.time{year=year, month=month, day=day, hour=hour, min=min})

					table.insert(data.site.posts, data.page)

					data.page.categoryDisplay = {}

					-- Populate category table
					for _, category in ipairs(data.page.categories) do
						if not data.site.categories[category] then
							data.site.categories[category] = {}
						end

						table.insert(data.site.categories[category], data.page)
						table.insert(data.page.categoryDisplay, { category = category, url = data.site.categoryMap[category]})
					end
				end

				lfs.mkdir("public/" .. data.page.url)
				writeFile("public/" .. data.page.url .. "index.html", lustache:render(templates[directory], data, templates))

				data.page = {}
			end
		end
	end
end

-- Render posts
lfs.mkdir("public")
render("posts")

-- Sort post archives
table.sort(data.site.posts, sortDates)

-- Renger pages
render("pages")

-- Feed
for loop=1, data.site.feedItems do
	data.site.feed[loop] = data.site.posts[loop]
end

writeFile("public/atom.xml", lustache:render(templates.atom, data, templates))
status("Rendering atom.xml")
data.page = {}

-- Categories
lfs.mkdir("public/categories")

for category in pairs(data.site.categories) do
	local categoryURL = data.site.categoryMap[category] .. "/"

	table.sort(data.site.categories[category], sortDates)

	data.page.title = category
	data.page.url = "categories/" .. categoryURL
	data.site.posts = data.site.categories[category]

	lfs.mkdir("public/categories/" .. categoryURL)
	writeFile("public/categories/" .. categoryURL .. "index.html", lustache:render(templates.categories, data, templates))
	status("Rendering " .. categoryURL)
end


src/motyl.rb
#!/usr/bin/env ruby
###############################################################################
#                                                                             #
# Motyl                                                                       #
# Copyright (c) 2016-2018, Frederic Cambus                                    #
# https://github.com/fcambus/motyl                                            #
#                                                                             #
# Created: 2016-02-16                                                         #
# Last Updated: 2018-03-18                                                    #
#                                                                             #
# Motyl is released under the BSD 2-Clause license.                           #
# See LICENSE file for details.                                               #
#                                                                             #
###############################################################################

require 'fileutils'
require 'kramdown'
require 'mustache'
require 'yaml'

# Load and process Markdown file
def markdown(path)
  return Kramdown::Document.new(File.read(path), syntax_highlighter: 'rouge').to_html
end

# Display status message
def status(message)
  puts('[' + Time.now.strftime('%X') + '] ' + message)
end

# Loading configuration
data = {
  'version' => 'Motyl 1.0.0',
  'updated' => Time.now.strftime('%Y-%m-%dT%XZ'),
  'site' => YAML.load_file('motyl.conf'),
  'posts' => [],
  'categories' => {}
}

# Loading templates
templates = {
  'categories' => File.read('themes/templates/categories.mustache'),
  'atom' => File.read('themes/templates/atom.mustache'),
  'pages' => File.read('themes/templates/page.mustache'),
  'posts' => File.read('themes/templates/post.mustache')
}

class Mustache
  self.template_path = 'themes/templates/'
end

def render(directory, templates, data)
  Dir.foreach(directory) do |file|
    next if file == '.' || file == '..'
    extension = File.extname(file)

    if extension == '.md'
      basename = File.basename(file, extension)
      data['page'] = YAML.load_file(directory + '/' + basename + '.yaml')
      data['page']['content'] = Mustache.render(markdown(directory + '/' + file), data)
      data['page']['url'] ||= basename + '/'

      status('Rendering ' + data['page']['url'])

      if directory == 'posts'
        data['page']['datetime'] = DateTime.parse(data['page']['date'])

        data['posts'].push(data['page'])

        data['page']['categoryDisplay'] = []

        # Populate category table
        data['page']['categories'].each do |category|
          data['categories'][category] ||= []
          data['categories'][category].push(data['page'])
          data['page']['categoryDisplay'].push('category' => category, 'url' => data['site']['categoryMap'][category])
        end
      end

      FileUtils.mkdir_p('public/' + data['page']['url'])
      File.write('public/' + data['page']['url'] + 'index.html', Mustache.render(templates[directory], data))

      data['page'] = {}
    end
  end
end

# Render posts
FileUtils.rm_rf('public')
render('posts', templates, data)

# Sort post archives
data['posts'].sort! { |a, b| b['date'] <=> a['date'] }

# Renger pages
render('pages', templates, data)

# Feed
data['feed'] = data['posts'][0..20]

File.write('public/atom.xml', Mustache.render(templates['atom'], data))
status('Rendering atom.xml')
data['page'] = {}

# Categories
data['categories'].keys.each do |category|
  category_url = data['site']['categoryMap'][category] + '/'

  data['categories'][category].sort! { |a, b| b['date'] <=> a['date'] }
  data['page']['title'] = category
  data['page']['url'] = 'categories/' + category_url
  data['posts'] = data['categories'][category]

  FileUtils.mkdir_p('public/categories/' + category_url)
  File.write('public/categories/' + category_url + 'index.html', Mustache.render(templates['categories'], data))
  status('Rendering ' + category_url)
end

# Copy static assets
status('Copying assets and static files')
FileUtils.cp_r('themes/fonts', 'public')
FileUtils.cp_r('themes/styles', 'public')
FileUtils.cp_r('static', 'public')
